Don't advise cloning when using relational operators

### Code

```Rust
#[derive(PartialEq, Eq, Clone)]
struct Foo {}

fn main() {
    let foo = Foo {};
    let bar = &Foo {};

    if foo == bar {
        println!("yay");
    }
}
```


### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:9:12
  |
9 |     if foo == bar {
  |        ---    ^^^ expected struct `Foo`, found `&Foo`
  |        |
  |        expected because this is `Foo`
  |
help: consider using clone here
  |
9 |     if foo == bar.clone() {
  |                  ++++++++
```


### Desired output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:9:12
  |
9 |     if foo == bar {
  |        ---    ^^^ expected struct `Foo`, found `&Foo`
  |        |
  |        expected because this is `Foo`
  |
help: consider dereferencing bar
  |
9 |     if foo == *bar {
  |               +
```


### Rationale and extra context

when using relational operators a clone is not necessary, and just dereferencing is more performant (for large types, obviously not in this case) and more readable.

### Other cases

_No response_

### Anything else?

_No response_

@rustbot modify labels +D-papercut

--- a/compiler/rustc_hir_typeck/src/demand.rs
+++ b/compiler/rustc_hir_typeck/src/demand.rs
@@ -1508,6 +1508,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         // FIXME(compiler-errors): We can actually do this if the checked_ty is
                         // `steps` layers of boxes, not just one, but this is easier and most likely.
                         || (checked_ty.is_box() && steps == 1)
+                        // We can always deref a binop that takes its arguments by ref.
+                        || matches!(
+                            self.tcx.hir().get_parent(expr.hir_id),
+                            hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Binary(op, ..), .. })
+                                if !op.node.is_by_value()
+                        )
                     {
                         let deref_kind = if checked_ty.is_box() {
                             "unboxing the value"

--- a/tests/ui/inference/deref-suggestion.rs
+++ b/tests/ui/inference/deref-suggestion.rs
@@ -72,4 +72,13 @@ fn main() {
     } else {
         &0
     };
+
+    #[derive(PartialEq, Eq)]
+    struct Foo;
+    let foo = Foo;
+    let bar = &Foo;
+
+    if foo == bar {
+    //~^ ERROR mismatched types
+    }
 }

--- a/tests/ui/inference/deref-suggestion.stderr
+++ b/tests/ui/inference/deref-suggestion.stderr
@@ -175,6 +175,19 @@ LL | ||     };
    |  |_____`if` and `else` have incompatible types
    |        expected `i32`, found `&{integer}`
 
-error: aborting due to 13 previous errors
+error[E0308]: mismatched types
+  --> $DIR/deref-suggestion.rs:81:15
+   |
+LL |     if foo == bar {
+   |        ---    ^^^ expected `Foo`, found `&Foo`
+   |        |
+   |        expected because this is `Foo`
+   |
+help: consider dereferencing the borrow
+   |
+LL |     if foo == *bar {
+   |               +
+
+error: aborting due to 14 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

